### PR TITLE
fix: Boundary condition error in VerifyStream

### PR DESCRIPTION
**Pull request by seebees**
Fixes: #507

Tests exist that push a single byte at a time.
This will hit every boundary condition.
However this is not complete enough.
State can be reset as the code transitions
from one boundary condition to another.

The prime example of this problem
the `VerifyStream`.
It has 4 boundary in the `_transform` function.
The body header, the body, the auth tag, and signature.
By sending 1 byte
as the code transitions from the body header
to the body
the code handling the exiting
of the body header has no way to interact
with the body,
because the 1 byte of data
completely isolates these code branches.

In this case the code was buffering
the body header in `frameBuffer`
but when exiting this section
and passing control to
the actual content section,
it used the `chunk` to form the tail.

When passing a single byte,
this tail would _always_ be empty,
because it would have been incorporated
into the frame header.

Adding tests to span
the boundary conditions
to verify both this fix
and that there are not other lurking.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: modules/decrypt-node/test/decrypt.test.ts [hunk 1]
```diff
@@ -59,33 +59,86 @@ describe('decrypt', () => {
...
+     * But there can be state that can get reset
+     * from a 1 byte push when the buffering
+     * is internal to the stream.
+     * So while 1 byte will hit the beginning
+     * boundary conditions,
+     * and the exiting boundary conditions,
+     * but it will never span any boundary conditions.
+     * The prime example of this is
+     * the `VerifyStream`.
+     * It has 4 boundary in the `_transform` function.
```
Comment (robin-aws) [inline modules/decrypt-node/test/decrypt.test.ts:73]:
```suggestion
     * It has 4 byte boundary in the `_transform` function.
```

?

File: modules/decrypt-node/test/decrypt.test.ts [hunk 2]
```diff
@@ -59,33 +59,86 @@ describe('decrypt', () => {
...
+     * Doing this randomly is a bad idea.
+     * Tests that only fail _sometimes_
+     * are bad tests.
+     * It is too easy to try again,
+     * and when everything passes just move on.
+     * And again, doing every permutation
+     * is too expensive at this time.
+     */
+    const results = await Promise.all(
+      [
```
Comment (robin-aws) [inline modules/decrypt-node/test/decrypt.test.ts:104]: nit: there must be a more compact way to get this value :)

File: modules/decrypt-node/test/decrypt.test.ts [hunk 3]
```diff
@@ -59,33 +59,86 @@ describe('decrypt', () => {
...
     )
 
-    expect(messageHeader.suiteId).to.equal(
-      AlgorithmSuiteIdentifier.ALG_AES256_GCM_IV12_TAG16_HKDF_SHA384_ECDSA_P384
-    )
-    expect(messageHeader.encryptionContext).to.deep.equal(
-      fixtures.encryptionContext()
-    )
-    expect(test.toString('base64')).to.equal(fixtures.base64Plaintext())
+    results.map(({ plaintext: test, messageHeader }) => {
```
Comment (robin-aws) [inline modules/decrypt-node/test/decrypt.test.ts:133]: Do we have a good option for more direct support for parameterized tests? It'll really suck to debug which size triggered a test failure if there's no output to tell which size was being used.

**Review by robin-aws — CHANGES_REQUESTED** [review]

File: modules/decrypt-node/test/decrypt.test.ts (hunk 3 above)
Comment (seebees) [reply modules/decrypt-node/test/decrypt.test.ts:133]: Not really no :(

File: modules/decrypt-node/test/decrypt.test.ts (hunk 2 above)
Comment (seebees) [reply modules/decrypt-node/test/decrypt.test.ts:104]: This is actually in response to your question below.
Since there is no nice parameterized test,
it is easiest to just comment out the ones you don't care about
to run the test and debug the issue...
